File: cbn/cbn/custom/production_plan.py
# Copyright (c) 2025, DAS and Contributors
# License: GNU General Public License v3. See license.txt

import frappe

def update_batch_manufacture(self, method=None):
    if not self.is_batch_manufacture:
        return

    frappe.set_value("Batch Manufacture", self.batch_manufacture, "status", "Used" if self.docstatus == 1 else "Empty")


def add_conversion_batch_manufacture(self, method=None):
    if not self.is_batch_manufacture:
        return

    for item in self.get("po_items"):
        item_parent = frappe.get_cached_value("Item", item.item_code, "custom_item_parent")
        if item_parent != self.item_master:
            frappe.throw("Item {} does not belong to item {}". format(item.code, self.batch_manufacture))

        add_conversion = "add_conversion"
        try:
            frappe.db.savepoint(add_conversion)
            batch_manufacture = frappe.get_doc("Batch Manufacture", self.batch_manufacture)
            batch_manufacture.append("item_conversion", {
                "item_code": item.item_code
            })
            batch_manufacture.flags.ignore_permissions = 1
            batch_manufacture.save()
        except frappe.UniqueValidationError:
            frappe.message_log.pop()
            frappe.db.rollback(save_point=add_conversion)

File: cbn/cbn/custom/production_plan.js
// Copyright (c) 2025, DAS and contributors
// For license information, please see license.txt

frappe.ui.form.on("Production Plan", {
	refresh(frm) {
		var doc = frm.doc;

        frm.set_query("batch_manufacture", function (doc) {
            if(!doc.item_master){
                frappe.throw("Select Production Item First")
            }

			if(!doc.posting_date){
                frappe.throw("Select Date First")
            }

			return {
				query: "cbn.controllers.queries.batch_manufacture_query",
				filters: {
                    item_code: doc.item_master,
                    status: "Empty",
					date: doc.posting_date,
				},
			};
		});
    },
	planned_start_date(frm){
		frm.doc.po_items.forEach((value) => {
			value.planned_start_date = frm.doc.planned_start_date
		});
		refresh_field("po_items");
	}
})

frappe.ui.form.on("Production Plan Item", {
	po_items_add(frm, cdt, cdn){
		var item = locals[cdt][cdn]

		if(frm.doc.planned_start_date){
			item.planned_start_date = frm.doc.planned_start_date
		}

		refresh_field("po_items");
	}
});